Integer overflow when creating threads

#### System Information
- Ogre Version: 1.9.0 (issue still present on master and 
- Operating System / Platform: Ubuntu 18.04 / AArch64 
- RenderSystem: Mesa / llvmpipe

#### Detailled description
In DefaultWorkQueue::startup(), there  is a loop used to initialise the threads - one per processor engine (core/thread) available on system:

```
for (uint8 i = 0; i < mWorkerThreadCount; ++i)                           
{                                                                        
    OGRE_THREAD_CREATE(t, *mWorkerFunc);                                 
    mWorkers.push_back(t);                                               
} 
```

Whilst **mWorkerThreadCount** is a size_t, uint8 was used for the index variable **i** which in this case can hold only 255 cores/threads. On my 256-threads system this becomes an infinite loop. After creating thousands of threads boost runs out of resource and the application/ogre crashes.

Issue can be seen since 1.7.0 until today's master (67852f4).

#### Ogre.log
The log is huge and most of the output is from the application using Ogre.
Here are the main points related to Ogre:

**There are thousands of these entries:**
```
[ INFO] [1548077974.061998103]: DefaultWorkQueue('Root') initialising on thread ffff85465000.
```

**Then eventually:**
```
rviz::RenderSystem: error creating render window: boost::thread_resource_error: Resource temporarily unavailable
```

**Then the application/Ogre trips over:**
```
[ WARN] [1548077974.062236977]: OGRE EXCEPTION(2:InvalidParametersException): Window with name 'OgreWindow(1)' already exists in GLRenderSystem::_createRenderWindow at /build/ogre-1.9-jztnFF/ogre-1.9-1.9.0+dfsg1/RenderSystems/GL/src/OgreGLRenderSystem.cpp (line 1057)
rviz::RenderSystem: error creating render window: OGRE EXCEPTION(2:InvalidParametersException): Window with name 'OgreWindow(1)' already exists in GLRenderSystem::_createRenderWindow at /build/ogre-1.9-jztnFF/ogre-1.9-1.9.0+dfsg1/RenderSystems/GL/src/OgreGLRenderSystem.cpp (line 1057)
...
[ERROR] [1548077974.070900242]: Unable to create the rendering window after 100 tries.
```
And finally the application segfaults.

--- a/OgreMain/src/OgreRoot.cpp
+++ b/OgreMain/src/OgreRoot.cpp
@@ -145,9 +145,10 @@ namespace Ogre {
         // never process responses in main thread for longer than 10ms by default
         defaultQ->setResponseProcessingTimeLimit(10);
         // match threads to hardware
-        unsigned threadCount = OGRE_THREAD_HARDWARE_CONCURRENCY;
-        if (!threadCount)
-            threadCount = 1;
+        int threadCount = OGRE_THREAD_HARDWARE_CONCURRENCY;
+        // but clamp it at 2 by default - we dont scale much beyond that currently
+        // yet it helps on android where it needlessly burns CPU
+        threadCount = Math::Clamp(threadCount, 1, 2);
         defaultQ->setWorkerThreadCount(threadCount);
 
         // only allow workers to access rendersystem if threadsupport is 1

--- a/OgreMain/src/Threading/OgreDefaultWorkQueueStandard.cpp
+++ b/OgreMain/src/Threading/OgreDefaultWorkQueueStandard.cpp
@@ -64,7 +64,7 @@ namespace Ogre
             Root::getSingleton().getRenderSystem()->preExtraThreadsStarted();
 
         mNumThreadsRegisteredWithRS = 0;
-        for (uint8 i = 0; i < mWorkerThreadCount; ++i)
+        for (size_t i = 0; i < mWorkerThreadCount; ++i)
         {
             OGRE_THREAD_CREATE(t, *mWorkerFunc);
             mWorkers.push_back(t);

